Sources/SwiftFusion/Nonlinear/NonlinearFactorGraph.swift

```diff
@@ -0,0 +1,114 @@
+import TensorFlow
+
+public struct NonlinearFactorGraph {
+  var factor1s: [Factor1] = []
+  var factor2s: [Factor2] = []
+
+  var factorCount: Int { factor1s.count + factor2s.count }
+
+  public init() {}
+
+  public mutating func add(
+    _ value1Index: Int, _ error: @escaping @differentiable (Pose2) -> Pose2, _ noise: NoiseModel
+  ) {
+    factor1s.append(Factor1(value1Index: value1Index, error: error, noise: noise))
+  }
+
+  public mutating func add(
+    _ value1Index: Int, _ value2Index: Int,
+    _ error: @escaping @differentiable (Pose2, Pose2) -> Pose2,
+    _ noise: NoiseModel
+  ) {
+    factor2s.append(Factor2(
+      value1Index: value1Index, value2Index: value2Index, error: error, noise: noise))
+  }
+
+  public func loss(at point: [Pose2]) -> Double {
+    factor1s.map { $0.noise.loss($0.error(point[$0.value1Index])) }.reduce(0, +) +
+      factor2s.map { $0.noise.loss($0.error(point[$0.value1Index], point[$0.value2Index])) }.reduce(0, +)
+  }
+
+  /// Returns an nx1 matrix (a vector) whose entries are the factors' errors at `point`.
+  public func errors(at point: [Pose2]) -> Tensor<Double> {
+    var vector = Tensor<Double>(zeros: [3 * factorCount, 1])
+    var currentFactorIndex: Int = 0
+    for factor in factor1s {
+      let error = factor.error(point[factor.value1Index])
+      vector[3 * currentFactorIndex, 0] = Tensor(error.rot.theta)
+      vector[3 * currentFactorIndex + 1, 0] = Tensor(error.t.x)
+      vector[3 * currentFactorIndex + 2, 0] = Tensor(error.t.y)
+      currentFactorIndex += 1
+    }
+    for factor in factor2s {
+      let error = factor.error(point[factor.value1Index], point[factor.value2Index])
+      vector[3 * currentFactorIndex, 0] = Tensor(error.rot.theta)
+      vector[3 * currentFactorIndex + 1, 0] = Tensor(error.t.x)
+      vector[3 * currentFactorIndex + 2, 0] = Tensor(error.t.y)
+      currentFactorIndex += 1
+    }
+    return vector
+  }
+
+  /// Returns a matrix representing a linear approximation of changes in factor errors as a
+  /// function of changes in `point`.
+  ///
+  /// Note: Currently returns a dense representation of the matrix, even though the matrix values
+  /// are actually very sparse.
+  public func linearize(at point: [Pose2]) -> Tensor<Double> {
```

Comment:
Yeah, dense will never be needed :-) So, I think it is as with above: just call linearize on each factor to return a JacobianFactor, which would be the GTSAM way. *But* with Fan we are debating whether we even need that: we just need vjp and jvp defined for each factor to implement CG. I think we need to be TDD here :-)